src/ops/constant/typecheck.rs

```diff
@@ -87,132 +54,3 @@ pub(crate) fn check_int_fits_in_width(
-        }
-        Container::Tuple(elems) => Container::Tuple(map_row(*elems, f)),
-        Container::Sum(variants) => Container::Sum(map_row(*variants, f)),
-        Container::Array(elem, sz) => Container::Array(Box::new(f(*elem)), sz),
-        Container::Alias(s) => Container::Alias(s),
-        Container::Opaque(custom) => Container::Opaque(custom),
-    }
-}
-
-/// Typecheck a constant value
-pub(super) fn typecheck_const(typ: &ClassicType, val: &ConstValue) -> Result<(), ConstTypeError> {
-    match (typ, val) {
-        (ClassicType::Hashable(HashableType::Int(exp_width)), ConstValue::Int(value)) => {
-            check_int_fits_in_width(*value, *exp_width).map_err(ConstTypeError::Int)
-        }
-        (ClassicType::F64, ConstValue::F64(_)) => Ok(()),
-        (ty @ ClassicType::Container(c), tm) => match (c, tm) {
-            (Container::Tuple(row), ConstValue::Tuple(xs)) => {
-                if row.len() != xs.len() {
-                    return Err(ConstTypeError::TupleWrongLength);
-                }
-                for (ty, tm) in row.iter().zip(xs.iter()) {
-                    typecheck_const(ty, tm)?
-                }
-                Ok(())
-            }
-            (Container::Tuple(_), _) => Err(ConstTypeError::ValueCheckFail(ty.clone(), tm.clone())),
-            (Container::Sum(row), ConstValue::Sum(tag, val)) => {
-                if let Some(ty) = row.get(*tag) {
-                    typecheck_const(ty, val.as_ref())
-                } else {
-                    Err(ConstTypeError::InvalidSumTag)
-                }
-            }
-            (Container::Sum(_), _) => Err(ConstTypeError::ValueCheckFail(ty.clone(), tm.clone())),
-            (Container::Opaque(ty), ConstValue::Opaque((ty_act, _val))) => {
-                if ty_act != ty {
-                    return Err(ConstTypeError::TypeMismatch(
-                        ty.clone().into(),
-                        ty_act.clone().into(),
-                    ));
-                }
-                Ok(())
-            }
-            _ => Err(ConstTypeError::Unimplemented(ty.clone())),
-        },
-        (ClassicType::Hashable(HashableType::Container(c)), tm) => {
-            // Here we deliberately build malformed Container-of-Hashable types
-            // (rather than Hashable-of-Container) in order to reuse logic above
-            typecheck_const(
-                &ClassicType::Container(map_vals(c.clone(), &ClassicType::Hashable)),
-                tm,
-            )
-        }
-        (ty @ ClassicType::Graph(_), _) => Err(ConstTypeError::Unimplemented(ty.clone())),
-        (ty @ ClassicType::Hashable(HashableType::String), _) => {
-            Err(ConstTypeError::Unimplemented(ty.clone()))
-        }
-        (ClassicType::Hashable(HashableType::Variable(_)), _) => {
-            Err(ConstTypeError::ConstCantBeVar)
-        }
-        (ty, _) => Err(ConstTypeError::ValueCheckFail(ty.clone(), val.clone())),
-    }
-}
-
-#[cfg(test)]
```

Comment:
These have gone to ops/constant.rs. I could write some HashableValue-specific ones in `values.rs`...